core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/util/DatabaseUtil.java

```diff
@@ -571,71 +574,117 @@ public static Connection getDBConnection(DataSource dataSource) throws SQLExcept
     }
 
     public static void closeConnection(Connection dbConnection) {
-
-        if (dbConnection != null) {
-            try {
-                dbConnection.close();
-            } catch (SQLException e) {
-                log.error("Database error. Could not close statement. Continuing with others. - " + e.getMessage(), e);
-            }
-        }
+        close(dbConnection);
     }
 
-    private static void closeResultSet(ResultSet rs) {
+    // Although using generics is a bit more cryptic to the uninitiated, this method ensures type safety
+    // and flexibility without the horrible cast conversions in
+    // favor of the usage of interfaces.
+    // Only java.sql types can be used as parameters.
+    //
+    // Only  java.sql object types intersect both Wrapper and AutoCloseable interfaces
+    // Types that are allowed as arguments: CachedRowSet, CallableStatement, Connection, DatabaseMetaData, DataSource,
+    // FilteredRowSet, JdbcRowSet, JoinRowSet, ParameterMetaData,
+    // PreparedStatement, ResultSet, ResultSetMetaData, RowSet, RowSetMetaData, Statement, SyncResolver, WebRowSet
+    private static <AutoClosableWrapper extends AutoCloseable & Wrapper> void close(AutoClosableWrapper dbObject) {
+
+        if (dbObject != null) {
 
-        if (rs != null) {
             try {
-                rs.close();
+                dbObject.close();
+                dbObject = null;
+            } catch (SQLRecoverableException ex) {
+                handleSQLRecoverableException(dbObject, ex, true);
```

Comment:
IMO, handling connection close if it fails should be up to the connection pool implementation. Retying for a second time does not feel right. Can you clarify the rationale behind this?